byterun/unix.c

```diff
@@ -347,28 +350,50 @@ CAMLexport int caml_read_directory(char * dirname, struct ext_table * contents)
 
 /* Recover executable name from /proc/self/exe if possible */
 
-#ifdef __linux__
-
-int caml_executable_name(char * name, int name_len)
+char * caml_executable_name(void)
 {
-  int retcode;
+#if defined(__linux__)
+  int namelen, retcode;
+  char * name;
   struct stat st;
 
-  retcode = readlink("/proc/self/exe", name, name_len);
-  if (retcode == -1 || retcode >= name_len) return -1;
+  /* lstat("/proc/self/exe") returns st_size == 0 so we cannot use it
+     to determine the size of the buffer.  Instead, we guess and adjust. */
+  namelen = 256;
+  while (1) {
+    name = caml_stat_alloc(namelen + 1);
+    retcode = readlink("/proc/self/exe", name, namelen);
+    if (retcode == -1) { caml_stat_free(name); return NULL; }
+    if (retcode <= namelen) break;
```

Comment:
Is this check right? If `readlink` has to truncate its response, won't `retcode == namelen`?

Reply:
Yes, but it can very well happen that more bytes than necessary were allocated, so no truncation is necessary, in which case you can have `retcode < namelen`.

Reply:
Right, but it seems like the only time that we want to break out of the loop is when `retcode < namelen` -- if `retcode == namelen` that's a sign that we want to try again with a longer name buffer.

Reply:
Yes, sounds right.  In fact, wouldn't it be even better to pass `namelen + 1` as third parameter of `readlink` and leave the comparison as it is ?

Reply:
Yeah, that is even better.

Reply:
Yes, the exit test should be `retcode < namelen`.

No, the length parameter of `readlink` must not be `namelen + 1` because we need room to add the terminating 0 that `readlink` doesn't add.

Reply:
@xavierleroy given the check `retcode <= namelen`, we would still have room for the null byte, as we'd still resize if `readlink` filled the entire buffer. And we could avoid reallocating when the length of the executable is exactly `namelen`.

(Though perhaps the additional mental cycles spent convincing oneself that this is safe is not worth saving one allocation approximately 0% of the time.)

Reply:
Fixed on trunk, commit e49e6ce, and cherry-picked to 4.05, commit 44c67eb.

Reply:
But if we leave the comparison as it is and we break out of the loop it means that `retcode <= namelen` so that there is space for the terminating 0, right?  Using `retcode < namelen` as exit test seems slightly less efficient because there could be some cases where `retcode == namelen` but not due to truncation.

Reply:
OK, you're both right.  I think the clearer code is to have `namelen` as the size of the buffer and as the 3rd parameter to `readlink`, and to break when `retcode < namelen`, which indeed leaves room for the final zero.  Pushing this improved fix in a minute...

Reply:
Commits 7a315bd on trunk and b5610e3.

Reply:
Oh yeah. That's a lot more clear.